Newsletters: Add subcribe modal setting (behind a feature flag)

edanzer opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to this issue: https://github.com/orgs/Automattic/projects/657/views/1?pane=issue&itemId=26582538

## Proposed Changes

This PR adds a new setting to enable a subscriber modal for Newsletter sites. It is still a draft intended to collect feedback.

Caveats:
   - The subscriber modal itself is not yet implemented. This PR will have no effect without that.
   - The setting is not being persisted a backend option. The settings page is using some non-obvious magic to do that and I still need to figure that part out. As is, if your refresh the page, the value always starts as false / off.
     - Question: Given that we will want to promote the newsletter feature to non-newsletter sites, I wonder if we should expose this to all sites, not just newsletter sites. 
   
<img width="1322" alt="subscribe setting" src="https://github.com/Automattic/wp-calypso/assets/21228350/2ab785ea-d860-410d-9a91-cdfa1ac5f612">

## Testing Instructions

1. Create or use a Newsletter site.  (`/setup/newsletter`)
2. Check out this branch and run yarn and yarn start if needed. 
3. Go to http://calypso.localhost:3000/settings/reading/YOURDOMAIN, scroll down to the Newsletter setting section. 
4. Confirm that the new section and new setting exists, and that the toggle works. 
5. Try loading same page on a non-newsletter site and confirm that setting does NOT show. 

Note that in production you'll have to use the feature flag in the URL: `?flags=newsletter/subscribe-modal`

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

edanzer committed:
Add subcribe modal setting

edanzer committed:
Remove period on setting label

edanzer committed:
Only show setting on newsletter sites

simison commented:
@crisbusquets FYI in case any design/copy input.

I did not know we have "Newsletter" section in settings, neat. :-) That's added when I was out on sabbatical I guess.

crisbusquets commented:
> I did not know we have "Newsletter" section in settings, neat. :-)

Me neither 😅

Thanks for the ping, I'll test this asap!

crisbusquets commented:
> Question: Given that we will want to promote the newsletter feature to non-newsletter sites, I wonder if we should expose this to all sites, not just newsletter sites.

I checked a regular site I created a while ago and `Settings > Reading` has the `Newsletter` block. I'd also add this new toggle there.

There are two dots at the end of the sentence, I'd remove the one that's with the link. And maybe the text should be `Edit the modal`.

<img width="742" alt="Screenshot 2023-05-24 at 11 45 03" src="https://github.com/Automattic/wp-calypso/assets/25607244/99f85b16-3fc6-4c73-9d10-8cfc65eccd8b">

Thanks @edanzer !

simison commented:
@crisbusquets We could make the label tiny bit more generic and ship other similar features and modals later on without adding more settings:

<img width="686" alt="image" src="https://github.com/Automattic/wp-calypso/assets/87168/321f4a90-d43b-4e37-b5b6-15fd6677ae78">

Is it worth starting with more broad copy now, or rather change it later?

crisbusquets commented:
> Is it worth starting with more broad copy now, or rather change it later?

If we have other modals/subscribe prompts in the future, we'll probably need different pages/modals to edit them individually. And I'm not sure if anyone will want to have 2-3 different prompts activated at the same time. I'd give our users the chance to choose.

edanzer committed:
Make link to subscribe modal template dynamic

edanzer committed:
Update modal option to match jetpack

edanzer committed:
Undo changes to yarn lock

edanzer committed:
Update option name to match Jetpack

edanzer committed:
Tweak subscribe modal link

edanzer commented:
Hey all - I've completed the [companion Jetpack PR](https://github.com/Automattic/jetpack/pull/31393) that goes along with this one (see that PR for testing instructions for the entire feature). 

I've updated this based on updates in the Jetpack PR, fixed the subscribe modal link, rebased, and re-tested this.

One the question of adding this for other site types, not just newsletter... I like the idea. If we do, we also need to update the companion Jetpack PR, which only adds the modal itself under specific condition (must be either newsletter site or have lettre theme activated, and must have subscribe modal enabled). 

Given that we're pushing the main code to Jetpack (released as a plugin via WP.org to millions of sites, with slower process to fix issues), and given that the modal is potentially intrusive on the front end, I may prefer to start by releasing the feature in a limited way on newsletter sites only. That way if there are any issues, the impact is limited. Once we 'soft launched' it for a bit on Newsletter sites, we can enable the feature for all sites. 

That sound OK?

simison commented on client/my-sites/site-settings/reading-newsletter-settings/SubscribeModalSetting.tsx:
```diff
@@ -0,0 +1,73 @@
+							link: (
+								<a
+									// TODO: add url to modal template
+									href={ localizeUrl( subscribeModalEditorUrl ) }
```
Not sure why we'd need `localizeUrl` here, the editor should be localized and the function is used more for support links etc. Removed in https://github.com/Automattic/wp-calypso/pull/77115/commits/f361b26299e6e87d3d824f9d8ba19eed949cedf3

simison committed:
Add feature flag

simison committed:
Remove unnessessary

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8350740

Thank you @edanzer for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8350740).
